File: rust/src/inference/nightly_trainer.rs
//! nightly_trainer.rs
//!
//! Ported from `CircleAI.Inference/NightlyAdapterTrainer.cs` plus the
//! `LoRAAdapterManager` surface from `MnnInteropRtFeatures.cs`.
//!
//! (Phase D3) Periodically drains the [`super::feedback_queue::IFeedbackTrainingQueue`],
//! runs LoRA gradient steps against the current model handle, saves the adapter,
//! and applies it. Idle-and-charging gating is host-supplied via a predicate.
//!
//! The native MNN training seam is injected behind [`ILoRAAdapterManager`]; the
//! background timer loop (`IHostedService.StartAsync` / `Task.Delay`) is host
//! plumbing, so the core drain-and-train pass is exposed as
//! [`NightlyAdapterTrainer::run_once`] — which the C# also exposes publicly for
//! manual triggering. [`InMemoryLoRAAdapterManager`] is a deterministic default
//! that computes a real (token-overlap) loss and records saved/applied adapters.

use std::sync::Arc;
use std::sync::Mutex;

use super::feedback_queue::{IFeedbackTrainingQueue, TrainingSample};

/// Outcome of one [`ILoRAAdapterManager::train_step`].
#[derive(Debug, Clone, PartialEq)]
pub enum TrainStepError {
    /// The native MNN binary was compiled without training support (mirrors the
    /// C# `NotSupportedException` from rc == -12). The trainer re-queues the
    /// batch and bails when it sees this.
    TrainingNotSupported,
    /// Any other training failure (mirrors the C# `InvalidOperationException`).
    Failed(String),
}

/// The LoRA-adapter native surface: run a gradient step, persist, apply.
/// Sync port of the training-relevant `LoRAAdapterManager` members.
pub trait ILoRAAdapterManager {
    /// Run one gradient-descent step on the LoRA adapter weights. Returns the
    /// scalar loss for the batch, or a [`TrainStepError`].
    fn train_step(
        &self,
        input_tokens: &[i32],
        target_tokens: &[i32],
        learning_rate: f32,
        lora_rank: i32,
    ) -> Result<f32, TrainStepError>;

    /// Persist the current LoRA adapter weights to `adapter_path`.
    fn save_adapter(&self, adapter_path: &str) -> Result<(), TrainStepError>;

    /// Apply a previously-saved adapter from `adapter_path` to the live model.
    fn apply(&self, adapter_path: &str) -> Result<(), TrainStepError>;
}

/// Deterministic in-memory [`ILoRAAdapterManager`]. Computes a real loss from
/// the input/target token overlap (lower overlap → higher loss), and records
/// the paths it saved / applied so tests can assert the trainer wired them.
#[derive(Debug, Default)]
pub struct InMemoryLoRAAdapterManager {
    training_supported: bool,
    saved: Mutex<Vec<String>>,
    applied: Mutex<Vec<String>>,
    step_count: Mutex<usize>,
}

impl InMemoryLoRAAdapterManager {
    /// A manager whose native training path is available.
    pub fn new() -> Self {
        Self {
            training_supported: true,
            saved: Mutex::new(Vec::new()),
            applied: Mutex::new(Vec::new()),
            step_count: Mutex::new(0),
        }
    }

    /// A manager that reports training is unavailable (`train_step` returns
    /// [`TrainStepError::TrainingNotSupported`]).
    pub fn without_training() -> Self {
        Self {
            training_supported: false,
            ..Self::new()
        }
    }

    /// Adapter paths passed to [`ILoRAAdapterManager::save_adapter`].
    pub fn saved_adapters(&self) -> Vec<String> {
        self.saved.lock().unwrap().clone()
    }

    /// Adapter paths passed to [`ILoRAAdapterManager::apply`].
    pub fn applied_adapters(&self) -> Vec<String> {
        self.applied.lock().unwrap().clone()
    }

    /// Number of successful training steps run.
    pub fn step_count(&self) -> usize {
        *self.step_count.lock().unwrap()
    }
}

impl ILoRAAdapterManager for InMemoryLoRAAdapterManager {
    fn train_step(
        &self,
        input_tokens: &[i32],
        target_tokens: &[i32],
        learning_rate: f32,
        lora_rank: i32,
    ) -> Result<f32, TrainStepError> {
        if input_tokens.is_empty() {
            return Err(TrainStepError::Failed("inputTokens required".into()));
        }
        if target_tokens.is_empty() {
            return Err(TrainStepError::Failed("targetTokens required".into()));
        }
        if learning_rate <= 0.0 {
            return Err(TrainStepError::Failed("learningRate out of range".into()));
        }
        if lora_rank <= 0 {
            return Err(TrainStepError::Failed("loraRank out of range".into()));
        }
        if !self.training_supported {
            return Err(TrainStepError::TrainingNotSupported);
        }

        // Deterministic loss: 1 - overlap ratio of target tokens present in the
        // input, scaled by learning rate. Real, reproducible, and monotone in
        // how well the target already matches the input.
        let matches = target_tokens
            .iter()
            .filter(|t| input_tokens.contains(t))
            .count();
        let overlap = matches as f32 / target_tokens.len() as f32;
        let loss = (1.0 - overlap) * (1.0 + learning_rate);

        *self.step_count.lock().unwrap() += 1;
        Ok(loss)
    }

    fn save_adapter(&self, adapter_path: &str) -> Result<(), TrainStepError> {
        if adapter_path.trim().is_empty() {
            return Err(TrainStepError::Failed("adapterPath required".into()));
        }
        self.saved.lock().unwrap().push(adapter_path.to_string());
        Ok(())
    }

    fn apply(&self, adapter_path: &str) -> Result<(), TrainStepError> {
        if adapter_path.trim().is_empty() {
            return Err(TrainStepError::Failed("adapterPath required".into()));
        }
        self.applied.lock().unwrap().push(adapter_path.to_string());
        Ok(())
    }
}

/// A tokeniser converting text → int IDs. The default char-level tokeniser maps
/// each UTF-16 code unit to its value, matching the C# `CharTokenizer` fallback.
pub type Tokenizer = Arc<dyn Fn(&str) -> Vec<i32> + Send + Sync>;

/// A gate deciding whether the trainer should fire now (battery / charging /
/// idle), matching the C# `Func<bool> ShouldFireNow`.
pub type ShouldFireNow = Arc<dyn Fn() -> bool + Send + Sync>;

/// Options for the nightly adapter trainer. Mirrors `NightlyAdapterTrainerOptions`.
#[derive(Clone)]
pub struct NightlyAdapterTrainerOptions {
    /// Minimum samples to bother training. Skip otherwise. Default 16.
    pub min_batch_size: i32,
    /// Cap per run so a backlog can't lock the device. Default 256.
    pub max_samples_per_run: i32,
    /// Adam-style LR for the LoRA adapter parameters. Default 1e-4.
    pub learning_rate: f32,
    /// Rank of the LoRA decomposition. Default 8.
    pub lora_rank: i32,
    /// Where to persist the trained adapter file. Default `circleai-lora.mnn`.
    pub adapter_path: String,
    /// Optional gate (battery/charging/idle). `None` = always fire.
    pub should_fire_now: Option<ShouldFireNow>,
    /// Tokeniser. `None` uses the char-level fallback.
    pub tokenizer: Option<Tokenizer>,
}

impl Default for NightlyAdapterTrainerOptions {
    fn default() -> Self {
        Self {
            min_batch_size: 16,
            max_samples_per_run: 256,
            learning_rate: 1e-4,
            lora_rank: 8,
            adapter_path: "circleai-lora.mnn".to_string(),
            should_fire_now: None,
            tokenizer: None,
        }
    }
}

/// Result of one [`NightlyAdapterTrainer::run_once`] pass.
#[derive(Debug, Clone, PartialEq)]
pub struct RunOnceResult {
    /// Number of successful training steps.
    pub steps: usize,
    /// Average loss across the steps (0.0 when `steps == 0`).
    pub avg_loss: f32,
    /// `true` when the batch was skipped because `pending < min_batch_size`.
    pub skipped_below_min: bool,
    /// `true` when training was unavailable and the batch was re-queued.
    pub requeued_unsupported: bool,
    /// `true` when the adapter was saved + applied.
    pub adapter_committed: bool,
}

/// (Phase D3) Drains the feedback queue and trains a LoRA adapter in one pass.
pub struct NightlyAdapterTrainer<Q: IFeedbackTrainingQueue, A: ILoRAAdapterManager> {
    queue: Q,
    adapter: A,
    opts: NightlyAdapterTrainerOptions,
}

impl<Q: IFeedbackTrainingQueue, A: ILoRAAdapterManager> NightlyAdapterTrainer<Q, A> {
    /// Constructs the trainer over a feedback queue + adapter manager.
    pub fn new(queue: Q, adapter: A, opts: NightlyAdapterTrainerOptions) -> Self {
        Self {
            queue,
            adapter,
            opts,
        }
    }

    /// Access the underlying adapter manager (test helper).
    pub fn adapter(&self) -> &A {
        &self.adapter
    }

    /// Access the underlying queue (test helper).
    pub fn queue(&self) -> &Q {
        &self.queue
    }

    /// `true` when the host gate (if any) permits firing now.
    pub fn should_fire_now(&self) -> bool {
        match &self.opts.should_fire_now {
            Some(g) => g(),
            None => true,
        }
    }

    /// (Phase D3) Drain + train in one pass. Reproduces the C# `RunOnceAsync`:
    /// skip when `pending < min_batch_size`; drain up to `max_samples_per_run`;
    /// tokenise input = user text, target = preferred text when polarity >= 0
    /// else assistant text; accumulate loss over train steps; on
    /// [`TrainStepError::TrainingNotSupported`] re-queue the whole batch and
    /// bail; on `steps > 0` save + apply the adapter.
    pub fn run_once(&self) -> RunOnceResult {
        let mut result = RunOnceResult {
            steps: 0,
            avg_loss: 0.0,
            skipped_below_min: false,
            requeued_unsupported: false,
            adapter_committed: false,
        };

        if (self.queue.pending() as i32) < self.opts.min_batch_size {
            result.skipped_below_min = true;
            return result;
        }

        let samples = match self.queue.drain(self.opts.max_samples_per_run) {
            Ok(s) => s,
            Err(_) => return result,
        };
        if samples.is_empty() {
            return result;
        }

        let mut total_loss = 0f32;
        let mut step_count = 0usize;

        for sample in &samples {
            let input = self.tokenize(&sample.user_text);
            let target = if sample.polarity >= 0 {
                self.tokenize(&sample.preferred_text)
            } else {
                self.tokenize(&sample.assistant_text)
            };
            if input.is_empty() || target.is_empty() {
                continue;
            }

            match self.adapter.train_step(
                &input,
                &target,
                self.opts.learning_rate,
                self.opts.lora_rank,
            ) {
                Ok(loss) => {
                    total_loss += loss;
                    step_count += 1;
                }
                Err(TrainStepError::TrainingNotSupported) => {
                    // Native MNN not built with training — re-queue and bail.
                    for s in &samples {
                        let _ = self.queue.enqueue(s.clone());
                    }
                    result.requeued_unsupported = true;
                    return result;
                }
                Err(TrainStepError::Failed(_)) => {
                    // Step failed for this sample — skip it (mirrors the C#
                    // per-sample warn-and-continue).
                }
            }
        }

        if step_count > 0 {
            let saved = self.adapter.save_adapter(&self.opts.adapter_path);
            let applied = self.adapter.apply(&self.opts.adapter_path);
            result.adapter_committed = saved.is_ok() && applied.is_ok();
        }

        result.steps = step_count;
        result.avg_loss = if step_count > 0 {
            total_loss / step_count as f32
        } else {
            0.0
        };
        result
    }

    fn tokenize(&self, text: &str) -> Vec<i32> {
        match &self.opts.tokenizer {
            Some(t) => t(text),
            None => char_tokenizer(text),
        }
    }
}

/// (Phase D3) Char-level tokeniser fallback — every UTF-16 code unit becomes its
/// value, matching the C# `CharTokenizer`.
pub fn char_tokenizer(text: &str) -> Vec<i32> {
    if text.is_empty() {
        return Vec::new();
    }
    text.encode_utf16().map(|u| u as i32).collect()
}

/// Convenience helper to build a re-usable [`TrainingSample`] for tests/hosts.
pub fn sample_now(
    user: impl Into<String>,
    assistant: impl Into<String>,
    preferred: impl Into<String>,
    polarity: i32,
) -> TrainingSample {
    TrainingSample::new(user, assistant, preferred, polarity, chrono::Utc::now())
}

File: python/src/circle_ai/inference/nightly_trainer.py
"""Nightly LoRA adapter trainer (Phase D3) + LoRA adapter manager (RT-10).

Ports:
  * ``CircleAI.Inference.LoRAAdapterManager`` — apply / read / unapply / train /
    save on a loaded model. The C# manager P/Invokes ``mnnbridge`` LoRA calls;
    Python injects the native seam behind :class:`ILoRANative`, with a
    deterministic in-memory default that computes a reproducible loss and
    persists adapter weights to disk so train/save/apply round-trip.
  * ``CircleAI.Inference.NightlyAdapterTrainerOptions`` + ``NightlyAdapterTrainer``
    — periodically drains the feedback queue, runs LoRA gradient steps, saves
    the adapter, and applies it. The idle-and-charging gate is host-supplied.
    Python drops the ``IHostedService`` background loop (no timer host here) and
    exposes :meth:`run_once_async` as the drain+train pass — the same public
    entry point the C# host triggers manually.
"""
from __future__ import annotations

import json
import os
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import Callable, List, Optional

from .feedback_training import IFeedbackTrainingQueue, TrainingSample

__all__ = [
    "TrainingNotSupportedError",
    "ILoRANative",
    "InMemoryLoRANative",
    "LoRAAdapterManager",
    "NightlyAdapterTrainerOptions",
    "NightlyAdapterTrainer",
    "char_tokenizer",
]


class TrainingNotSupportedError(RuntimeError):
    """Raised when the native runtime was built without training support.

    Mirrors the C# ``NotSupportedException`` thrown by ``TrainStep`` when the
    bridge returns ``MNNBRIDGE_ERR_TRAINING_DISABLED`` (-12). The trainer
    re-queues the batch and skips the run when this is raised.
    """


class ILoRANative(ABC):
    """Injected native seam over the mnnbridge LoRA ABI. Mirrors the P/Invoke
    surface ``LoRAAdapterManager`` binds to.

    ``train_step`` returns a raw status code (0 success, -12 training disabled,
    other < 0 = error) plus the loss via the return tuple.
    """

    @abstractmethod
    def apply(self, handle: object, adapter_path: str) -> int: ...

    @abstractmethod
    def unapply(self, handle: object) -> int: ...

    @abstractmethod
    def current_adapter(self, handle: object) -> Optional[str]: ...

    @abstractmethod
    def train_step(
        self,
        handle: object,
        input_tokens: List[int],
        target_tokens: List[int],
        learning_rate: float,
        lora_rank: int,
    ) -> tuple[int, float]: ...

    @abstractmethod
    def save_lora(self, handle: object, adapter_path: str) -> int: ...


class InMemoryLoRANative(ILoRANative):
    """Deterministic in-memory stand-in for the mnnbridge LoRA ABI.

    Maintains per-handle "adapter weights" (a running accumulator over training
    steps) so that :meth:`train_step` returns a reproducible, monotonically
    decreasing loss, :meth:`save_lora` persists the weights, :meth:`apply`
    records the last-applied path, and :meth:`current_adapter` reads it back.

    Set ``training_enabled=False`` to simulate a bridge built without
    ``MNN_BUILD_TRAIN`` (train_step returns -12).
    """

    __slots__ = ("_training_enabled", "_weights", "_applied", "_steps")

    def __init__(self, training_enabled: bool = True) -> None:
        self._training_enabled = training_enabled
        self._weights: dict[int, float] = {}
        self._applied: dict[int, str] = {}
        self._steps: dict[int, int] = {}

    def apply(self, handle: object, adapter_path: str) -> int:
        if handle is None:
            return -1
        self._applied[id(handle)] = adapter_path
        return 0

    def unapply(self, handle: object) -> int:
        if handle is None:
            return -1
        self._applied.pop(id(handle), None)
        return 0

    def current_adapter(self, handle: object) -> Optional[str]:
        if handle is None:
            return None
        return self._applied.get(id(handle))

    def train_step(
        self,
        handle: object,
        input_tokens: List[int],
        target_tokens: List[int],
        learning_rate: float,
        lora_rank: int,
    ) -> tuple[int, float]:
        if not self._training_enabled:
            return (-12, 0.0)
        if handle is None:
            return (-1, 0.0)
        hid = id(handle)
        step = self._steps.get(hid, 0) + 1
        self._steps[hid] = step
        # Deterministic "loss": shrinks as steps accumulate, scaled by the
        # input/target token magnitude. Purely for reproducible bookkeeping.
        base = (sum(input_tokens) + sum(target_tokens)) % 97 + 1
        loss = base / (step + lora_rank)
        self._weights[hid] = self._weights.get(hid, 0.0) + learning_rate * loss
        return (0, float(loss))

    def save_lora(self, handle: object, adapter_path: str) -> int:
        if handle is None:
            return -1
        d = os.path.dirname(adapter_path)
        if d:
            os.makedirs(d, exist_ok=True)
        weight = self._weights.get(id(handle), 0.0)
        steps = self._steps.get(id(handle), 0)
        with open(adapter_path, "w", encoding="utf-8") as fh:
            json.dump({"weight": weight, "steps": steps}, fh)
        return 0


class LoRAAdapterManager:
    """RT-10 LoRA adapter manager. Port of ``CircleAI.Inference.LoRAAdapterManager``.

    Wraps a model ``handle`` (opaque) and an injected :class:`ILoRANative`
    (defaults to :class:`InMemoryLoRANative`). Methods mirror the C# surface
    and raise on native error codes.
    """

    __slots__ = ("_handle", "_native")

    def __init__(self, handle: object, native: Optional[ILoRANative] = None) -> None:
        self._handle = handle
        self._native = native if native is not None else InMemoryLoRANative()

    def apply(self, adapter_path: str) -> None:
        if not adapter_path or not adapter_path.strip():
            raise ValueError("adapter_path required")
        if not (os.path.isfile(adapter_path) or os.path.isdir(adapter_path)):
            raise FileNotFoundError(f"LoRA adapter not found: {adapter_path}")
        r = self._native.apply(self._handle, adapter_path)
        if r != 0:
            raise RuntimeError(f"apply_lora failed: {r}")

    def unapply(self) -> None:
        r = self._native.unapply(self._handle)
        if r != 0:
            raise RuntimeError(f"unapply_lora failed: {r}")

    def current_adapter(self) -> Optional[str]:
        return self._native.current_adapter(self._handle)

    def train_step(
        self,
        input_tokens: List[int],
        target_tokens: List[int],
        learning_rate: float = 1e-4,
        lora_rank: int = 8,
    ) -> float:
        """Run one gradient-descent step; return the scalar loss. Raises
        :class:`TrainingNotSupportedError` when the native runtime lacks
        training support (mirrors the C# ``NotSupportedException`` on -12).
        """
        if not input_tokens:
            raise ValueError("input_tokens required")
        if not target_tokens:
            raise ValueError("target_tokens required")
        if learning_rate <= 0:
            raise ValueError("learning_rate must be > 0")
        if lora_rank <= 0:
            raise ValueError("lora_rank must be > 0")
        rc, loss = self._native.train_step(
            self._handle, input_tokens, target_tokens, learning_rate, lora_rank
        )
        if rc == -12:
            raise TrainingNotSupportedError(
                "native runtime was compiled without training support. "
                "Rebuild with training enabled for on-device LoRA fine-tuning."
            )
        if rc != 0:
            raise RuntimeError(f"train_lora_step failed: {rc}")
        return loss

    def save_adapter(self, adapter_path: str) -> None:
        if not adapter_path or not adapter_path.strip():
            raise ValueError("adapter_path required")
        d = os.path.dirname(adapter_path)
        if d:
            os.makedirs(d, exist_ok=True)
        rc = self._native.save_lora(self._handle, adapter_path)
        if rc != 0:
            raise RuntimeError(f"save_lora failed: {rc}")


def char_tokenizer(text: str) -> List[int]:
    """Char-level tokenizer fallback — each char becomes its code-point value.

    Port of ``NightlyAdapterTrainer.CharTokenizer``.
    """
    if not text:
        return []
    return [ord(c) for c in text]


@dataclass(frozen=True, slots=True)
class NightlyAdapterTrainerOptions:
    """Options for the nightly trainer. Mirrors
    ``CircleAI.Inference.NightlyAdapterTrainerOptions``.

    * ``min_batch_size`` — minimum samples to bother training; skip otherwise.
    * ``max_samples_per_run`` — cap per run so a backlog can't lock the device.
    * ``learning_rate`` — LR for the LoRA adapter parameters.
    * ``lora_rank`` — rank of the LoRA decomposition.
    * ``adapter_path`` — where to persist the trained adapter file.
    * ``interval_seconds`` — how often to check whether to train (default 6h).
    * ``should_fire_now`` — optional gate (battery/charging/idle); default fires.
    * ``tokenizer`` — text -> int IDs; falls back to :func:`char_tokenizer`.
    """

    min_batch_size: int = 16
    max_samples_per_run: int = 256
    learning_rate: float = 1e-4
    lora_rank: int = 8
    adapter_path: str = "circleai-lora.mnn"
    interval_seconds: float = 6 * 60 * 60
    should_fire_now: Optional[Callable[[], bool]] = None
    tokenizer: Optional[Callable[[str], List[int]]] = None


@dataclass(frozen=True, slots=True)
class NightlyRunResult:
    """Outcome of a single :meth:`NightlyAdapterTrainer.run_once_async`.

    Not present in C# (there the method is ``void``), added here so callers can
    assert on what happened without scraping logs.
    """

    trained_steps: int
    average_loss: float
    skipped_reason: Optional[str] = None
    saved: bool = False


class NightlyAdapterTrainer:
    """Drains the feedback queue and trains a LoRA adapter. Port of
    ``CircleAI.Inference.NightlyAdapterTrainer``.

    The C# type is an ``IHostedService`` with a timer loop; the drain+train
    logic lives in ``RunOnceAsync``. Python ports ``run_once_async`` faithfully
    and drops the timer host (a caller schedules it). ``should_fire_now`` is
    honoured inside :meth:`run_once_async` so the gate is testable.
    """

    __slots__ = ("_queue", "_adapter", "_opts")

    def __init__(
        self,
        queue: IFeedbackTrainingQueue,
        adapter: LoRAAdapterManager,
        opts: NightlyAdapterTrainerOptions,
    ) -> None:
        if queue is None:
            raise ValueError("queue is required")
        if adapter is None:
            raise ValueError("adapter is required")
        if opts is None:
            raise ValueError("opts is required")
        self._queue = queue
        self._adapter = adapter
        self._opts = opts

    async def run_once_async(self, ct: object = None) -> NightlyRunResult:
        """Drain + train in one pass. Mirrors ``RunOnceAsync``.

        * Skips when fewer than ``min_batch_size`` samples are pending.
        * Tokenizes each sample (target = preferred for polarity>=0, else the
          assistant text), runs a train step, and averages the loss.
        * On :class:`TrainingNotSupportedError`, re-queues the whole batch and
          bails (matching the C# re-queue-and-return).
        * Saves + applies the adapter when any step ran.
        """
        # Optional host gate.
        if self._opts.should_fire_now is not None and not self._opts.should_fire_now():
            return NightlyRunResult(0, 0.0, skipped_reason="gate_closed")

        if self._queue.pending < self._opts.min_batch_size:
            return NightlyRunResult(
                0,
                0.0,
                skipped_reason=f"pending {self._queue.pending} < min {self._opts.min_batch_size}",
            )

        samples = await self._queue.drain_async(self._opts.max_samples_per_run, ct)
        if len(samples) == 0:
            return NightlyRunResult(0, 0.0, skipped_reason="drained_empty")

        tokenizer = self._opts.tokenizer or char_tokenizer
        total_loss = 0.0
        step_count = 0
        for sample in samples:
            try:
                inp = tokenizer(sample.user_text)
                target = tokenizer(
                    sample.preferred_text if sample.polarity >= 0 else sample.assistant_text
                )
                if len(inp) == 0 or len(target) == 0:
                    continue
                loss = self._adapter.train_step(
                    inp, target, self._opts.learning_rate, self._opts.lora_rank
                )
                total_loss += loss
                step_count += 1
            except TrainingNotSupportedError:
                # Native training unavailable — re-queue and bail.
                for s in samples:
                    await self._queue.enqueue_async(s, ct)
                return NightlyRunResult(0, 0.0, skipped_reason="training_disabled")
            except Exception:
                # Per-sample step failure — skip this sample, keep going.
                continue

        saved = False
        avg_loss = 0.0
        if step_count > 0:
            avg_loss = total_loss / step_count
            try:
                self._adapter.save_adapter(self._opts.adapter_path)
                self._adapter.apply(self._opts.adapter_path)
                saved = True
            except Exception:
                saved = False

        return NightlyRunResult(step_count, avg_loss, saved=saved)
